File: internal/session/cookie.go
// Package session provides hardened cookie management for ZITADEL sessions.
// All session cookies are HMAC-SHA256 signed, HttpOnly, SameSite=Lax, and
// Secure (when not on localhost). Cookie format: base64(token.hmac_signature).
package session

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"fmt"
	"net/http"
	"strings"
)

const (
	// secureCookieName uses the __Host- prefix for maximum browser protection:
	// forces Secure=true, Path=/, no Domain attribute, can't be set from subdomains.
	secureCookieName = "__Host-zitadel_session"

	// devCookieName is used in local development (no HTTPS).
	devCookieName = "__zitadel_session"

	// MaxAge is the session cookie lifetime in seconds (24 hours).
	MaxAge = 86400
)

// CookieConfig holds the settings needed for cookie operations.
type CookieConfig struct {
	// Secrets is an ordered list of HMAC keys. The first key is used for signing;
	// all keys are tried for verification (enables zero-downtime key rotation).
	// If empty, a random key is generated and stored in-memory (dev mode).
	Secrets []string

	// Secure forces the Secure flag on cookies. Derived from ExternalDomain.
	Secure bool
}

// NewCookieConfig creates a CookieConfig from server settings.
// If secrets is empty, a random 32-byte key is generated (logged for debugging).
func NewCookieConfig(secrets []string, externalDomain string) *CookieConfig {
	secure := externalDomain != "" && externalDomain != "localhost" && externalDomain != "127.0.0.1"

	if len(secrets) == 0 {
		key := make([]byte, 32)
		rand.Read(key)
		secrets = []string{hex.EncodeToString(key)}
	}

	return &CookieConfig{
		Secrets: secrets,
		Secure:  secure,
	}
}

// CookieName returns the appropriate cookie name based on Secure mode.
func (c *CookieConfig) CookieName() string {
	if c.Secure {
		return secureCookieName
	}
	return devCookieName
}

// SetSessionCookie writes a hardened session cookie to the response.
// The token is HMAC-signed: cookie_value = base64(token + "." + hex(hmac)).
func SetSessionCookie(w http.ResponseWriter, token string, cfg *CookieConfig) {
	signed := sign(token, cfg.Secrets[0])

	http.SetCookie(w, &http.Cookie{
		Name:     cfg.CookieName(),
		Value:    signed,
		Path:     "/",
		MaxAge:   MaxAge,
		HttpOnly: true,
		Secure:   cfg.Secure,
		SameSite: http.SameSiteLaxMode,
	})
}

// ClearSessionCookie removes the session cookie.
func ClearSessionCookie(w http.ResponseWriter, cfg *CookieConfig) {
	http.SetCookie(w, &http.Cookie{
		Name:     cfg.CookieName(),
		Value:    "",
		Path:     "/",
		MaxAge:   -1,
		HttpOnly: true,
		Secure:   cfg.Secure,
		SameSite: http.SameSiteLaxMode,
	})
}

// ReadSessionCookie extracts and verifies the session token from the request.
// Returns the raw token (for DB lookup) and true if valid, or ("", false).
// For backward compatibility, also accepts unsigned tokens (legacy cookies).
func ReadSessionCookie(r *http.Request, cfg *CookieConfig) (string, bool) {
	// Try both cookie names (secure and dev) for transition periods.
	for _, name := range []string{secureCookieName, devCookieName} {
		cookie, err := r.Cookie(name)
		if err != nil || cookie.Value == "" {
			continue
		}

		// Try HMAC-signed format first.
		if token, ok := verify(cookie.Value, cfg.Secrets); ok {
			return token, true
		}

		// Backward compatibility: accept raw (unsigned) tokens.
		// This allows existing sessions to continue working after upgrade.
		if !strings.Contains(cookie.Value, ".") && len(cookie.Value) > 0 {
			return cookie.Value, true
		}
	}

	return "", false
}

// sign produces a signed cookie value: base64(token.hex(hmac)).
func sign(token, secret string) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(token))
	sig := hex.EncodeToString(mac.Sum(nil))
	payload := fmt.Sprintf("%s.%s", token, sig)
	return base64.RawURLEncoding.EncodeToString([]byte(payload))
}

// verify checks a signed cookie value against all secrets (key rotation).
// Returns the raw token and true if any key validates the signature.
func verify(cookieValue string, secrets []string) (string, bool) {
	decoded, err := base64.RawURLEncoding.DecodeString(cookieValue)
	if err != nil {
		return "", false
	}

	parts := strings.SplitN(string(decoded), ".", 2)
	if len(parts) != 2 {
		return "", false
	}

	token := parts[0]
	providedSig := parts[1]

	for _, secret := range secrets {
		mac := hmac.New(sha256.New, []byte(secret))
		mac.Write([]byte(token))
		expectedSig := hex.EncodeToString(mac.Sum(nil))
		if hmac.Equal([]byte(providedSig), []byte(expectedSig)) {
			return token, true
		}
	}

	return "", false
}

File: crates/zitadel-authn/src/cookie.rs
use base64::{Engine, engine::general_purpose::URL_SAFE_NO_PAD};
use hmac::{Hmac, Mac};
use sha2::Sha256;

type HmacSha256 = Hmac<Sha256>;

const SECURE_COOKIE_NAME: &str = "__Host-zitadel_session";
const DEV_COOKIE_NAME: &str = "__zitadel_session";
/// Default max age for session cookies (24h). Can be overridden via `SessionConfig.max_age_secs`.
pub const DEFAULT_MAX_AGE: i64 = 86400;

/// Cookie configuration with HMAC key rotation support.
#[derive(Clone)]
pub struct CookieConfig {
    /// Ordered list of HMAC keys. First signs, all verify.
    pub secrets: Vec<String>,
    /// Whether to use Secure flag (derived from external_domain).
    pub secure: bool,
    /// Maximum cookie/session age in seconds.
    pub max_age: i64,
}

impl CookieConfig {
    pub fn new(secrets: Vec<String>, external_domain: &str, force_insecure_cookies: bool) -> Self {
        Self::new_with_max_age(
            secrets,
            external_domain,
            force_insecure_cookies,
            DEFAULT_MAX_AGE,
        )
    }

    pub fn new_with_max_age(
        secrets: Vec<String>,
        external_domain: &str,
        force_insecure_cookies: bool,
        max_age: i64,
    ) -> Self {
        let secure = !force_insecure_cookies
            && !external_domain.is_empty()
            && external_domain != "localhost"
            && external_domain != "127.0.0.1";

        let secrets = if secrets.is_empty() {
            vec![zitadel_crypto::random_hex(32)]
        } else {
            secrets
        };

        Self {
            secrets,
            secure,
            max_age,
        }
    }

    pub fn cookie_name(&self) -> &str {
        if self.secure {
            SECURE_COOKIE_NAME
        } else {
            DEV_COOKIE_NAME
        }
    }

    /// All accepted cookie names (for reading during transition).
    pub fn all_cookie_names(&self) -> &[&str] {
        &[SECURE_COOKIE_NAME, DEV_COOKIE_NAME]
    }
}

/// Sign a token for cookie storage: base64url(token.hex(hmac)).
pub fn sign(token: &str, secret: &str) -> String {
    let mut mac =
        HmacSha256::new_from_slice(secret.as_bytes()).expect("HMAC accepts any key length");
    mac.update(token.as_bytes());
    let sig = hex::encode(mac.finalize().into_bytes());
    let payload = format!("{token}.{sig}");
    URL_SAFE_NO_PAD.encode(payload.as_bytes())
}

/// Verify a signed cookie value against all secrets (key rotation).
/// Returns the raw token if any key validates.
pub fn verify(cookie_value: &str, secrets: &[String]) -> Option<String> {
    let decoded = URL_SAFE_NO_PAD.decode(cookie_value).ok()?;
    let decoded = String::from_utf8(decoded).ok()?;
    let (token, provided_sig) = decoded.split_once('.')?;

    for secret in secrets {
        let mut mac =
            HmacSha256::new_from_slice(secret.as_bytes()).expect("HMAC accepts any key length");
        mac.update(token.as_bytes());
        let expected_sig = hex::encode(mac.finalize().into_bytes());

        // Constant-time comparison via hmac crate isn't needed here since
        // we're comparing hex strings, but let's be safe.
        if constant_time_eq(provided_sig.as_bytes(), expected_sig.as_bytes()) {
            return Some(token.to_string());
        }
    }

    None
}

fn constant_time_eq(a: &[u8], b: &[u8]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    let mut diff = 0u8;
    for (x, y) in a.iter().zip(b.iter()) {
        diff |= x ^ y;
    }
    diff == 0
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn sign_verify_roundtrip() {
        let secret = "test-secret-key-for-hmac";
        let token = "session-token-123";
        let signed = sign(token, secret);
        let verified = verify(&signed, &[secret.to_string()]);
        assert_eq!(verified, Some(token.to_string()));
    }

    #[test]
    fn verify_wrong_secret_fails() {
        let signed = sign("token", "secret1");
        let result = verify(&signed, &["secret2".to_string()]);
        assert_eq!(result, None);
    }

    #[test]
    fn verify_key_rotation() {
        let signed = sign("token", "old-key");
        // Verify against both old and new keys
        let result = verify(&signed, &["new-key".to_string(), "old-key".to_string()]);
        assert_eq!(result, Some("token".to_string()));
    }

    #[test]
    fn verify_garbage_returns_none() {
        assert_eq!(verify("not-base64!!!", &["secret".to_string()]), None);
        assert_eq!(verify("", &["secret".to_string()]), None);
    }

    #[test]
    fn cookie_config_dev() {
        let cfg = CookieConfig::new(vec![], "localhost", false);
        assert!(!cfg.secure);
        assert_eq!(cfg.cookie_name(), DEV_COOKIE_NAME);
        assert!(!cfg.secrets.is_empty()); // Random key generated
    }

    #[test]
    fn cookie_config_production() {
        let cfg = CookieConfig::new(vec!["my-secret".into()], "auth.example.com", false);
        assert!(cfg.secure);
        assert_eq!(cfg.cookie_name(), SECURE_COOKIE_NAME);
    }

    #[test]
    fn cookie_config_force_insecure() {
        let cfg = CookieConfig::new(vec!["my-secret".into()], "auth.example.com", true);
        assert!(!cfg.secure);
        assert_eq!(cfg.cookie_name(), DEV_COOKIE_NAME);
    }
}
